### PR TITLE
New module pcomclient

### PR DESCRIPTION
**Pull request by lmrosa**
PCOM client module that allows unauthenticated 
read and write different types of operands of Unitronics PLCs.

See https://unitronicsplc.com/Download/SoftwareUtilities/Unitronics%20PCOM%20Protocol.pdf

## Verification

List the steps needed to make sure this thing works

- [x] Start `msfconsole`
- [x] `use scanner/scada/pcomclient`
- [x] `set RHOST IP`
- [x] `run`
- [x] **Verify** the thing does what it should
- [x] **Verify** the thing does not do what it should not
- [x] **Document** the thing and how it works

### CODE REVIEW

File: modules/auxiliary/scanner/scada/pcomclient.rb [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        ans = write
+        if ans == nil
+          print_error("No answer from PLC")
+          return
+        end
+      end
+    else
+      print_error("Unknown action #{action.name}")
+    end
+    disconnect
```
Comment (wchen-r7) [inline modules/auxiliary/scanner/scada/pcomclient.rb]: Good practice on the cleanup with `disconnect`! However this is actually automatically taken care of in the `Msf::Exploit::Remote::Tcp` minx. So you can remove this line safely and just let the mixin do its thing :-)

File: modules/auxiliary/scanner/scada/pcomclient.rb [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        'DefaultAction' => 'READ'
+                     ))
+    register_options(
+      [
+        Opt::RPORT(20256),
+        OptInt.new('UNITID', [ false, 'Unit ID (0 - 127)', 0]),
+        OptInt.new('ADDRESS', [true, "PCOM memory address (0 - 65535)", 0]),
+        OptInt.new('LENGTH', [true, "Number of values to read (1 - 255) (read only)", 3]),
+        OptString.new('VALUES', [false, "Values to write (0 - 65535 each) (comma separated) (write only)"]),
+        OptEnum.new("OPERAND", [true, 'Operand type', "MI", ["Input", "Output", "SB", "MB", "MI", "SI", "ML", "SL"]]),
```
Comment (wchen-r7) [inline modules/auxiliary/scanner/scada/pcomclient.rb]: Very small detail, but that extra `,` isn't needed :-)

**Comment by wchen-r7** [comment]
Pretty well written module I'd say. Since this is a PLC, it would be difficult for us to test, @lmrosa do you still have this PLC? If so could you please provide a PCAP that demonstrates this module works? Thanks.

**Comment by lmrosa** [comment]
> Pretty well written module I'd say. Since this is a PLC, it would be difficult for us to test, @lmrosa do you still have this PLC? If so could you please provide a PCAP that demonstrates this module works? Thanks.

Thank you for the comments. Please refer to [pcaps](https://github.com/lmrosa/pcom-misc/tree/master/pcaps) to a list of PCOM samples I'm collecting.

File: modules/auxiliary/scanner/scada/pcomclient.rb (hunk 1 above)
Comment (lmrosa) [reply modules/auxiliary/scanner/scada/pcomclient.rb]: done

File: modules/auxiliary/scanner/scada/pcomclient.rb (hunk 2 above)
Comment (lmrosa) [reply modules/auxiliary/scanner/scada/pcomclient.rb]: done

**Comment by lmrosa** [comment]
@wchen-r7 am I missing something? Do you know why Travis CI build is failing? Is something related with imports I removed as you suggested ?

**Comment by bcoles** [comment]
It's unrelated to your changes. Something to do with `data/ysoserial_payloads.json`, which is from a different PR.

**Comment by bcoles** [comment]
@msjenkins-r7 test this please

**Comment by lmrosa** [comment]
@wchen-r7 @bcoles is the Travis CI pipeline fixed ?

**Comment by bcoles** [comment]
> 
> 
> @wchen-r7 @bcoles is the Travis CI pipeline fixed ?

Yes, Travis is working at the moment.

**Comment by wchen-r7** [comment]
I will cleanup and land this. Thanks!

**Comment by wchen-r7** [comment]
# Release Notes

PCOM client module that allows unauthenticated read and write different types of operands of Unitronics PLCs.

**Pull request merged**